engine/message/queue.go

```diff
@@ -0,0 +1,220 @@
+		sent:    make([]Message, 0, len(msgs)),
+		pending: make(map[notification.DestType][]Message),
+		now:     now,
+
+		serviceSent: make(map[string]time.Time),
+		userSent:    make(map[string]time.Time),
+		destSent:    make(map[notification.Dest]time.Time),
+	}
+
+	for _, m := range msgs {
+		if m.SentAt.IsZero() {
+			q.pending[m.Dest.Type] = append(q.pending[m.Dest.Type], m)
+			continue
+		}
+		q.addSent(m)
+	}
+
+	return q
+}
+func (q *queue) addSent(m Message) {
+	if m.SentAt.IsZero() {
+		m.SentAt = q.now
+	}
+
+	if t := q.serviceSent[m.ServiceID]; m.SentAt.After(t) {
+		q.serviceSent[m.ServiceID] = m.SentAt
+	}
+	if t := q.userSent[m.UserID]; m.SentAt.After(t) {
+		q.userSent[m.UserID] = m.SentAt
+	}
+	if t := q.destSent[m.Dest]; m.SentAt.After(t) {
+		q.destSent[m.Dest] = m.SentAt
+	}
+
+	q.sent = append(q.sent, m)
+}
+
+func (q *queue) userPriority(userA, userB string) (isLess, ok bool) {
+	sentA := q.userSent[userA]
+	sentB := q.userSent[userB]
+
+	if sentA.Equal(sentB) {
+		return false, false
+	}
+
+	return sentA.Before(sentB), true
+}
+
+func (q *queue) servicePriority(serviceA, serviceB string) (isLess, ok bool) {
+	sentA := q.serviceSent[serviceA]
+	sentB := q.serviceSent[serviceB]
+
+	if sentA.Equal(sentB) {
+		// neither has recieved a message
+		return false, false
+	}
+
+	return sentA.Before(sentB), true
+}
+
+// filterPending will delete messages from pending that are not eligible to be sent.
+func (q *queue) filterPending(destType notification.DestType) {
+	pending := q.pending[destType]
+	if len(pending) == 0 {
+		return
+	}
+
+	cutoffTime := q.now.Add(-cmCooldown)
+	filtered := pending[:0]
+	for _, p := range pending {
+		if q.destSent[p.Dest].After(cutoffTime) {
+			continue
+		}
+		filtered = append(filtered, p)
+	}
+
+	q.pending[destType] = filtered
+}
+
+// sortPending will re-sort the list of pending messages.
+func (q *queue) sortPending(destType notification.DestType) {
+	pending := q.pending[destType]
+	if len(pending) == 0 {
+		return
+	}
+
+	rand.Shuffle(len(pending), func(i, j int) { pending[i], pending[j] = pending[j], pending[i] })
+	sort.SliceStable(pending, func(i, j int) bool {
+		pi, pj := pending[i], pending[j]
+		if pi.CreatedAt.Equal(pj.CreatedAt) {
+			// keep existing order
+			return i < j
+		}
+		// sort by creation time (if timestamps are not equal)
+		return pi.CreatedAt.Before(pj.CreatedAt)
+	})
+	sort.SliceStable(pending, func(i, j int) bool {
+		pi, pj := pending[i], pending[j]
+
+		// First Alert to a service takes highest priority
+		piTypePriority := typePriority[pi.Type]
+		if pi.Type == TypeAlertNotification && q.serviceSent[pi.ServiceID].IsZero() {
+			piTypePriority = 0
+		}
+
+		pjTypePriority := typePriority[pj.Type]
+		if pj.Type == TypeAlertNotification && q.serviceSent[pj.ServiceID].IsZero() {
+			pjTypePriority = 0
+		}
+
+		if piTypePriority != pjTypePriority {
+			return piTypePriority < pjTypePriority
+		}
+
+		if isLess, ok := q.userPriority(pi.UserID, pj.UserID); ok {
```

Comment:
Should we return out with an error if not ok, or resort to random ordering?

Reply:
OK means "yes we have determined priority" but not an error scenario. Basically `ok` means `isLess` is valid. If `ok` is falls, then `isLess` should be ignored.

It's following the same pattern as type assertions and map existence checks:
```go
str, ok := interfaceVal.(string)

val, ok := someMap["value"]
```